docs: workshop_a_precision_convertor_journey: Initial version (#30)
analogdevicesinc/documentation: docs/learning/workshop_a_precision_converter_fpga_integration_journey/index.rst

> A Precision Converter FPGA Integration Journey
> ==============================================
> 
> Agenda
> ~~~~~~

Reviewer: ```suggestion
.. _workshop_precision_converter:

A Precision Converter FPGA Integration Journey
===============================================================================

Agenda
-------------------------------------------------------------------------------
```

All the lines should be 80 characters long.

1. Introduction
2. SPI Engine Architecture
3. What Is the Use Case?
4. System Build
5. System Evaluation

> 6. Conclusions

Reviewer: Should do automatic numbering, with #. and not manually write the numbers

Author: Done

Introduction
~~~~~~~~~~~~

**Customer Journey**

.. figure:: intro_customer_journey_1.png
   :align: center

Tools / Platforms are a customer choice, ADI ports to “current” development
environment/kit.

.. figure:: intro_customer_journey_2.png
   :align: center

> Maintenance is an ADI burden. Customer don’t start their designs at the same
> time, and want to use the latest and greatest.
> 
> **COS Reference Design “Donut Hole” Strategy**

Reviewer: ```suggestion
Maintenance is an ADI burden. Customer don't start their designs at the same
time, and want to use the latest and greatest.

**COS Reference Design "Donut Hole" Strategy**
```

- [ ] That type of single and double quotes messes with the pdf generation if I remember correctly, so please change them to these, in the whole page:

' 
and 
"

- [ ] Also, maybe it would be better if these would be made as subsections

Author: Fixed quotes

*Surround others processor, FPGA, micro ADI components*
*Reference design creates stickiness*
*Customer experiences low friction*
*Leverage ecosystems*
*Develop/participate in viable communities / ecosystems:*
- Linux kernel (1.3 Billion users)
- GitHub (40 Million users)
- Python (100 Million users)

> - MATLAB (1 Million users)

Reviewer: This is not rendered as separate lines, please review the page visually in the browser

Author: Done

.. figure:: intro_donut.png
   :align: center

**COS Full Stack High Level Overview**

.. figure:: intro_full_stack_overview.png
   :align: center

**COS Full Stack HDL Designs**

*Common HDL across all IIO reference designs*

> *Full stack reference Designs include JESD204 physical layers (XCVR), link layers, and transport layers.*
> - Termination to DMA via an AXI-STREAM or FIFO interface
> - Runs across different Intel and Xilinx carriers

Reviewer: ```suggestion
*Full stack reference Designs include JESD204 physical layers (XCVR), link layers, and transport layers.*

- Termination to DMA via an AXI-STREAM or FIFO interface
- Runs across different Intel and Xilinx carriers
```
Without that blank line, they're not rendered properly. This should be done everywhere where you want a list

- Arbitrates multiple command streams into a single CSE
- Arbitration is done at a SPI transaction level (a SYNC instruction must be used for end of transaction)
- Lower slave port always take precedence

.. figure:: spie_interconnect_ip.png
   :align: center

**SPI Engine Framework – Execution IP**

- Accepts commands on the control interface
- Generates low-level SPI signals on the SPI interface
- Active signal indicates if the block is busy processing commands

.. figure:: spie_execution_ip.png
   :align: center

**SPI Engine Framework – Command Stream Interfaces**

*4 AXI-Stream interfaces*
- CMD: Instructions
- SDO: SPI write data (MOSI)
- SDI: SPI read data (MISO)
- SYNC: Synchronization events

*Each streams follows standard AXI-Stream handshaking*
- Only the main AXI-stream signals are used (ready, valid, data)

**SPI Engine Framework – Software support**

- Introduces the concept of SPI offload
- Moves certain operations typically performed by the application processor to the SPI controller
- Interrupt offload capability
- Data offload capability
- SPI-Engine is one implementation of SPI offload
- ADI converter drivers can be used with any offload capable SPI controller

> 

Reviewer: ```suggestion
```

Author: removed

Resolution[bits]             16                     18
Sampling rate[KSPS]          15                     15 and 1330
Input signal frequency[kHz]  1                      1
Input signal amplitude[dBFS] -0.5                   -0.5
Supply voltage [V]           -2.5 and 5             -2.5 and 5
============================ ====================== =====================

**Features of the AD7984 converter**

- High performance
- Throughput: 1.33 MSPS
- True differential analog input range: ±VREF
- 0 V to VREF with VREF between 2.9 V to 5 V
- Zero latency architecture
- 18-bit resolution with no missing codes
- Dynamic range: 99.7 dB, VREF = 5 V
- SNR: 98.5 dB at fIN = 1 kHz, VREF = 5 V
- THD: −110.5 dB at fIN = 1 kHz, VREF = 5 V
- SINAD: 97.5 dB at fIN = 1 kHz, VREF = 5 V

**AD7984 SPI transfer Timing Diagram**

.. figure:: use_case_ad7984_transfer_diagram.png
   :align: center

**Timing parameters needed to configure the framework**

.. figure:: use_case_timing_parameters_ad7984.png
   :align: center

**HDL design block diagram**

.. figure:: use_case_hdl_bd.png
   :align: center

**HDL Framework instantiation**

- TCL Function header:

> proc spi_engine_create {{name "spi_engine"} {data_width 32} {async_spi_clk 1} {num_cs 1} {num_sdi 1} {sdi_delay 0} {echo_sclk 0}}

Reviewer: Should be in a "code-block" directive

Author: fixed

- Instantiation example for PulSAR ADC:

>   .. code:: tcl

Reviewer: ```suggestion
  .. code-block:: tcl
```
All "code" directives should be replaced with code-block

Author: fixed

     source $ad_hdl_dir/library/spi_engine/scripts/spi_engine.tcl4
     set data_width32
     set async_spi_clk1
     set num_cs1
     set num_sdi1
     set sdi_delay1
     set hier_spi_enginespi_pulsar_adc
     spi_engine_create $hier_spi_engine $data_width $async_spi_clk $num_cs $num_sdi $sdi_delay

> 1. Data_width - will set the width of the data bus / data line used by the SPI

Reviewer: With #.

   :align: center

**System Build - Power Supply**

.. figure:: system_build_power_supply.png
   :align: center

**System Build - Input Signal**

.. figure:: system_build_input_signal.png
   :align: center

**System Build - UART Configuration**

.. figure:: system_build_uart.png
   :align: center

**System Build – change host IP**

.. figure:: system_build_ip.png
   :align: center

**System Build - UART and Ethernet Testing**

.. figure:: system_build_uart_and_eth.png
   :align: center

Evaluate System
~~~~~~~~~~~~~~~

**System Evaluation – regular SPI trigger configuration**

.. figure:: system_evaluation_spi_trigger_1.png
   :align: center

.. figure:: system_evaluation_spi_trigger_2.png
   :align: center

**System Evaluation**

=========================================================== ====================== =====================
Steps                                                       Regular SPI controller SPI Engine controller
=========================================================== ====================== =====================
Connect IIO Oscilloscope and visualize the captured signals
Run the Python script on Cora z7s / Cygwin
Analyze results in VisualAnalog

> Compare the results

Reviewer 2: Is this correct? no checkbox is checked

Author: is correct, this is the first step - at this level we didn't do any step for now. If you look down, in the next table steps begin to be marked as completed

=========================================================== ====================== =====================

**System Evaluation - IIO Oscilloscope**

.. figure:: system_evaluation_iio_osc_1.png
   :align: center

.. figure:: system_evaluation_iio_osc_2.png
   :align: center

**System Evaluation - Logic Analyzer**

.. figure:: system_evaluation_m2k_1.png
   :align: center

.. figure:: system_evaluation_m2k_2.png
   :align: center

**System Evaluation**

=========================================================== ====================== =====================
Steps                                                       Regular SPI controller SPI Engine controller
=========================================================== ====================== =====================
Connect IIO Oscilloscope and visualize the captured signals           X

> Run the Python script on Cora z7s / Cygwin

Reviewer: - [ ] When referring to Cora, please do it like this:

```suggestion
Run the Python script on `Cora Z7S`_ / Cygwin
```

- [ ] And on the last line from the file, add this line:

.. _Cora Z7S: https://digilent.com/shop/cora-z7-zynq-7000-single-core-for-arm-fpga-soc-development

=========================================================== ====================== =====================
Connect IIO Oscilloscope and visualize the captured signals           X                     X
Run the Python script on Cora z7s / Cygwin                            X                     X
Analyze results in VisualAnalog                                       X                     X
Compare the results                                                   X                     X
=========================================================== ====================== =====================

Conclusions
~~~~~~~~~~~

1. A classic MCU can be used for converters that have the sampling rate up to 100kSPS.
2. Maximum performance, in terms of sampling rate, SNR, THD can be only achieved with an FPGA.
3. SPI Engine is a highly flexible and powerful open-source SPI controller
   framework which can interface a wide range of precision converters.
4. We are now familiar with the COS group open-source solution stack.

**Thank You!**

**Related Presentations**

- My customer uses a FPGA in his product. Now what?
- ADALM2000 in real life applications
- Just enough Software and HDL for High-Speed designs

> - Hardware and Software Tools for Precision Wideband Instrumentation

Reviewer 2: Maybe we could link to those, no?

Author: Cannot find this FTC presentations on System Level Docs, can you provide some hints where should I take the links?

Talking to Paul he is not aware that this presentations were ported...
